core/runtime/common/runtime_instances_pool.hpp

```diff
@@ -63,30 +65,39 @@ namespace kagome::runtime {
      * @return Module if any, nullopt otherwise
      */
     std::optional<std::shared_ptr<const Module>> getModule(
-        const RootHash &state);
+        const TrieHash &state);
 
     /**
      * @brief Puts new module into internal cache
      *
-     * @param state - runtime block, by its root hash
+     * @param state - storage hash of the block containing the code of the
+     * module
      * @param module - new module pointer
      */
-    void putModule(const RootHash &state, std::shared_ptr<Module> module);
+    void putModule(const TrieHash &state, std::shared_ptr<Module> module);
 
    private:
-    struct Entry {
+    struct InstancePool {
       std::shared_ptr<const Module> module;
       std::vector<std::shared_ptr<ModuleInstance>> instances;
 
-      outcome::result<std::shared_ptr<ModuleInstance>> instantiate(
-          std::unique_lock<std::mutex> &lock);
+      std::shared_ptr<ModuleInstance> instantiate(
+          std::unique_lock<std::shared_mutex> &lock);
     };
 
+    using CompilationResult =
+        outcome::result<std::shared_ptr<const Module>, CompilationError>;
+    CompilationResult tryCompileModule(const CodeHash &code_hash,
+                                       common::BufferView code_zstd);
+
     std::shared_ptr<ModuleFactory> module_factory_;
 
-    std::mutex mt_;
-    static constexpr size_t MODULES_CACHE_SIZE = 2;
-    Lru<common::Hash256, Entry> pools_;
+    std::shared_mutex pools_mtx_;
+    Lru<common::Hash256, InstancePool> pools_;
+
+    std::shared_mutex compiling_modules_mtx_;
```

Comment:
`shared_lock` is not used, use one `std::mutex` for both `pools_` and `compiling_modules_mtx_`

Reply:
Add a usage for shared pools_mtx_ in once place.